### ISSUE TITLE
Provide a way to do stateful filtering in CKEDITOR.htmlParser.filter

### ISSUE THREAD

**Issue by mmichaelis** [ckeditor/ckeditor-dev#871]
## Are you reporting a feature request or a bug?

Feature request

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

_none found_

## Provide detailed reproduction steps (if any)

I want to be able to have stateful rules in `CKEDITOR.htmlParser.filter`.

The actual use case:

In addition to the pastefromword filtering I want to process text-nodes with Symbol-font-family and replace Symbol characters by their corresponding entity or unicode replacement. To do so I would like to detect style changes and adapt text nodes accordingly later on.

The attached file shows the workaround by extending `CKEDITOR.htmlParser.filter`:

* [workaround-stateful-htmlfiltering.txt](https://github.com/ckeditor/ckeditor-dev/files/1280782/workaround-stateful-htmlfiltering.txt)

(It's ok to tell me that I am on the completely wrong track and should use another approach instead...).

### Expected result

All handler function in the rules-object receive the corresponding filter (or some other thing to track information during processing) as last argument.

### Actual result

Some handlers already receive the filter as last argument, as for example the elements-filter. From an API-user perspective it feels that the API is not straight in its behavior there.

## Other details

* Browser: all
* OS: all
* CKEditor version: 4.7.2
* Installed CKEditor plugins: pastefromword
* See Also:
    * [CKEDITOR.htmlParser.filter - CKEditor 4 Documentation](https://docs.ckeditor.com/#!/api/CKEDITOR.htmlParser.filter)

**Comment by msamsel** [comment]
Hi,

it depend how complicated operation you would like to made. If you wish to transform or modify pasted text to editor, simplest way would be to listen to `paste` event and transform data within.
Some example you can find in code pen below. Try to paste something there:
https://codepen.io/msamsel/pen/WEBPBB?editors=1010
I'm still not sure what exactly goal you have in this case, but maybe this approach will be suitable.

**Comment by mmichaelis** [comment]
Sorry for the late response. Meanwhile we track the state with a variable declared within the scope of the filter-rules. Nevertheless, it feels strange that some handlers receive 'this' as filter reference, while others don't. So I think it is at least a nice-to-have that all handlers have a similar contract regarding the parameters handed over.

What we try to achieve (and up to now we have to disable ACF):

When pasting from Word it is a common use case that symbols like &alpha;, &otimes;, etc. are not represented as Unicode symbols but as *normal* characters with [Symbol typeface](https://en.wikipedia.org/wiki/Symbol_(typeface)).

Our implementation is to listen to `afterPasteFromWord` (different to the provided example, we learned that it is better to hook in here) and to use `CKEDITOR.htmlParser.filter` within the handler directly using `filter.applyTo(fragment)` and later rebuild the HTML.

The rules contain rules for text, attribute `style` and `element` (any) and process as follows:
* element-handler tracks the current element-path in state-variable
* style-handler tracks the current font-family in state-variable
* text-handler does processing according to the stored state, thus if to do character replacement or not

As we found a solution meanwhile feel free to close this one as won't fix unless you think my arguments are reasonable that all rules should have some at least similar API regarding their function parameters.
